torch/csrc/jit/runtime/vararg_functions.cpp

```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
   push(stack, ss.str());
 }
 
+// IValue tags are intentionally private, so we need additional logic to cast
+// the IValue type to the specified format.
+std::string getFormattedArg(char key, const IValue& ival, int precision=6) {
+  //TODO: Implement precison-based formatting
+  std::stringstream ss;
+  bool added = false;
+  switch (key) {
+    case 'd':
+    case 'i': {
+      TORCH_CHECK(
+          ival.isScalar(),
+          "%",
+          key,
+          " format: A number is required, not ",
+          ival.tagKind());
+      if (ival.isDouble()) {
+        std::stringstream().swap(ss);
+        ss << static_cast<int>(ival.toDouble());
+        added = true;
+      }
+      break;
+    }
+    case 'e':
+    case 'E': {
+      TORCH_CHECK(
+        ival.isScalar(),
+        "%",
+        key,
+        " format: A number is required, not ",
+        ival.tagKind());
+      ss << std::setprecision(precision) << std::scientific;
+      if (key == 'E') {
+        ss << std::uppercase;
+      }
+      if (ival.isInt()) {
+        ss << static_cast<float>(ival.toInt());
+      } else {
+        ss << static_cast<float>(ival.toDouble());
+      }
+      added = true;
+      break;
+    }
+    case 'f':
+    case 'F': {
+      TORCH_CHECK(
+          ival.isScalar(),
+          "%",
+          key,
+          " format: A number is required, not ",
+          ival.tagKind());
+      ss << std::setprecision(precision) << std::fixed;
+      if (ival.isInt()) {
+        ss << static_cast<float>(ival.toInt());
+      } else {
+        ss << static_cast<float>(ival.toDouble());
+      }
+      added = true;
+      break;
+    }
+    case 'c': {
+      TORCH_CHECK(ival.isInt() || (ival.isString() && ival.toStringRef().length() == 1), "%c requires int or char");
+      if (ival.isInt()) {
+        ss << static_cast<char>(ival.toInt());
+        added = true;
+      }
+      break;
+    }
+  }
+  if (!added) {
+    ss << ival;
+  }
+  return ss.str();
+}
+
+void percentFormat(Stack& stack, size_t num_inputs) {
+  std::vector<char> specifiers = {'d', 'i', 'e', 'E', 'f', 'F', 's', 'c'};
```

Comment:
I think this can be made `const static`

Reply:
and make it into an `unordered_set` would make your check on line 132 much faster

Reply:
In fact, thinking more about it, this check isn't exactly necessary. The helper method needs to check validity of format char anyway, it can return whether the format is valid or not.